### PR TITLE
feat: ckTESTBTC

### PR DESCRIPTION
**Pull request by peterpeterparker**
# Motivation

Behind a flag provide ckTESTBTC in NNS-dapp on mainnet.

# Changes

- CKTESTBTC and canister ids introduced as a new universe
- ckBTC services and apis extended with caniser IDs as parameters
- ckbtc accounts stores -> icrc accounts stores
- `ENABLE_CKBTC_LEDGER` and `ENABLE_CKBTC_MINTER` renamed to `ENABLE_CKBTC` and `ENABLE_CKTESTBTC`
- addition map of ckBTC canisters added and pass to modal, that way modals which canister (index, minter) to use

# Screenshot

Flag: on

Capture d’écran 2023-02-22 à 10 27 44
Capture d’écran 2023-02-22 à 10 27 49
Capture d’écran 2023-02-22 à 10 27 51
Capture d’écran 2023-02-22 à 10 27 57
Capture d’écran 2023-02-22 à 10 28 04
Capture d’écran 2023-02-22 à 10 28 09
Capture d’écran 2023-02-22 à 10 28 11

Flag off:

Capture d’écran 2023-02-22 à 10 29 04

### CODE REVIEW

File: dfx.json [hunk 1]
```diff
@@ -237,8 +237,8 @@
           "ENABLE_SNS_2": false,
           "ENABLE_SNS_VOTING": false,
           "ENABLE_SNS_AGGREGATOR": false,
-          "ENABLE_CKBTC_LEDGER": true,
-          "ENABLE_CKBTC_MINTER": false
+          "ENABLE_CKBTC": true,
+          "ENABLE_CKTESTBTC": false
```
Comment (lmuntaner) [inline dfx.json:241]: This change could be it's own PR

File: frontend/src/lib/api/ckbtc-index.api.ts [hunk 2]
```diff
@@ -4,24 +4,24 @@ import type {
...
-import { CKBTC_INDEX_CANISTER_ID } from "$lib/constants/canister-ids.constants";
 import { HOST } from "$lib/constants/environment.constants";
 import { logWithTimestamp } from "$lib/utils/dev.utils";
 import type { HttpAgent, Identity } from "@dfinity/agent";
 import { IcrcIndexCanister } from "@dfinity/ledger";
+import type { Principal } from "@dfinity/principal";
 
 export const getCkBTCTransactions = async ({
   identity,
+  indexCanisterId: canisterId,
```
Comment (lmuntaner) [inline frontend/src/lib/api/ckbtc-index.api.ts:15]: Adding the parameter, could be it's own PR

File: frontend/src/lib/components/accounts/CkBTCTransactionsList.svelte [hunk 3]
```diff
@@ -4,13 +4,16 @@
...
-  import { CKBTC_UNIVERSE_CANISTER_ID } from "$lib/constants/canister-ids.constants";
   import {
     getSortedTransactionsFromStore,
     isIcrcTransactionsCompleted,
   } from "$lib/utils/icrc-transactions.utils";
   import IcrcTransactionsList from "$lib/components/accounts/IcrcTransactionsList.svelte";
+  import type { UniverseCanisterId } from "$lib/types/universe";
+  import type { CkBTCAdditionalCanisters } from "$lib/types/ckbtc-canisters";
 
+  export let canisters: CkBTCAdditionalCanisters;
```
Comment (lmuntaner) [inline frontend/src/lib/components/accounts/CkBTCTransactionsList.svelte:15]: You could have passed the `indexCanisterId` directly, right?

Any reason you preferred this way?

File: frontend/src/lib/components/universe/UniverseLogo.svelte [hunk 4]
```diff
@@ -20,13 +24,20 @@
...
-  $: logo = summary?.metadata.logo ?? (ckBTC ? CKBTC_LOGO : IC_LOGO);
+  $: logo =
+    summary?.metadata.logo ??
+    (ckTESTBTC ? CKTESTBTC_LOGO : ckBTC ? CKBTC_LOGO : IC_LOGO);
 
   let title: string;
   $: title =
     summary?.metadata.name !== undefined
       ? `${summary?.metadata.name} ${$i18n.sns_launchpad.project_logo}`
+      : ckTESTBTC
```
Comment (lmuntaner) [inline frontend/src/lib/components/universe/UniverseLogo.svelte:39]: This is already hard to read. Could we move this to a util or use a map?

File: frontend/src/lib/constants/ckbtc-additional-canister-ids.constants.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  CKBTC_MINTER_CANISTER_ID,
+  CKBTC_UNIVERSE_CANISTER_ID,
+  CKTESTBTC_INDEX_CANISTER_ID,
+  CKTESTBTC_MINTER_CANISTER_ID,
+  CKTESTBTC_UNIVERSE_CANISTER_ID,
+} from "$lib/constants/ckbtc-canister-ids.constants";
+import type { CkBTCAdditionalCanisters } from "$lib/types/ckbtc-canisters";
+import type { UniverseCanisterIdText } from "$lib/types/universe";
+
+export const CKBTC_ADDITIONAL_CANISTERS: Record<
```
Comment (lmuntaner) [inline frontend/src/lib/constants/ckbtc-additional-canister-ids.constants.ts:12]: 👏  I like the idea of grouping them

File: frontend/src/lib/derived/ckbtc-universes.derived.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+export const CKBTC_UNIVERSE: Universe = {
+  canisterId: CKBTC_UNIVERSE_CANISTER_ID.toText(),
+};
+
+export const CKTESTBTC_UNIVERSE: Universe = {
+  canisterId: CKTESTBTC_UNIVERSE_CANISTER_ID.toText(),
+};
+
+export const ckBTCUniversesStore = derived<
```
Comment (lmuntaner) [inline frontend/src/lib/derived/ckbtc-universes.derived.ts:20]: This could have been its own PR that uses just on universe but prepares to handle multiple

File: frontend/src/lib/derived/universes-tokens.derived.ts [hunk 7]
```diff
@@ -18,23 +20,36 @@ export const nnsTokenStore = derived<
 
 export const ckBTCTokenStore = derived<
```
Comment (lmuntaner) [inline frontend/src/lib/derived/universes-tokens.derived.ts:21]: Should we consider doing an icrcTokenStore?

File: frontend/src/lib/derived/universes-tokens.derived.ts [hunk 8]
```diff
@@ -18,23 +20,36 @@ export const nnsTokenStore = derived<
...
-);
+  Record<UniverseCanisterIdText, TokensStoreUniverseData | undefined>
+>([tokensStore], ([$tokensStore]) => ({
+  [CKBTC_UNIVERSE_CANISTER_ID.toText()]:
+    $tokensStore[CKBTC_UNIVERSE_CANISTER_ID.toText()],
+  [CKTESTBTC_UNIVERSE_CANISTER_ID.toText()]:
+    $tokensStore[CKTESTBTC_UNIVERSE_CANISTER_ID.toText()],
+}));
 
 export const ckBTCTokenFeeStore = derived<
```
Comment (lmuntaner) [inline frontend/src/lib/derived/universes-tokens.derived.ts:31]: Similar, does it make sense to do a `icrcTokenFeeStore`?

Or move the fee to the token data that is added along the metadata?

File: frontend/src/lib/stores/ckbtc-accounts.store.ts [hunk 9]
```diff
@@ -1,35 +0,0 @@
-import type { Account } from "$lib/types/account";
```
Comment (lmuntaner) [inline frontend/src/lib/stores/ckbtc-accounts.store.ts:1]: To use icrcAccountsStore instead of this could have been its own PR also

File: frontend/src/lib/stores/icrc-accounts.store.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    accounts: IcrcAccounts;
+    universeId: UniverseCanisterId;
+  }) => void;
+  reset: () => void;
+}
+
+/**
+ * A store that contains the ckBTC and other Icrc compatible accounts.
+ */
+export const initIcrcAccountsStore = (): IcrcAccountsStore => {
```
Comment (lmuntaner) [inline frontend/src/lib/stores/icrc-accounts.store.ts:27]: We could move snsAccountsStore here! I like it!

Or you could have renamed snsAccountsStore to icrcAccountsStore and use that one.

File: frontend/src/tests/lib/components/accounts/CkBTCAccountsFooter.spec.ts [hunk 11]
```diff
@@ -3,18 +3,28 @@
...
 describe("CkBTCAccountsFooter", () => {
+  beforeAll(() => {
+    page.mock({
+      data: { universe: CKTESTBTC_UNIVERSE_CANISTER_ID.toText() },
+      routeId: AppPath.Accounts,
+    });
+  });
+
   afterEach(() => {
     jest.clearAllMocks();
```
Comment (lmuntaner) [inline frontend/src/tests/lib/components/accounts/CkBTCAccountsFooter.spec.ts:25]: Can you move these to the `beforeEach` as agreed in the testing best practices?

File: frontend/src/lib/derived/selected-universe.derived.ts [hunk 12]
```diff
@@ -75,3 +78,12 @@ export const selectedUniverseStore: Readable<Universe> = derived(
       ({ canisterId }) => canisterId === $selectedUniverseIdStore.toText()
     ) ?? NNS_UNIVERSE
 );
+
+export const selectedCkBTCUniverseIdStore = derived<
```
Comment (lmuntaner) [inline frontend/src/lib/derived/selected-universe.derived.ts:82]: I didn't see a test for this one, did I miss it?

**Review by lmuntaner — COMMENTED** [review]
Why doesn't the screenshots use the new test logo?

Added a few comments.

I believe this could have been split into multiple smaller PRs. It's hard to review such a big PR and be confident.

For next time! Thanks!

**Comment by peterpeterparker** [comment]
> I believe this could have been split into multiple smaller PRs. It's hard to review such a big PR and be confident.

In the absolute, yes agree but, it was known that this would be open for at least two-three weeks so to some extension, I got other things to do that maintaining twenty PRs.

Let's try to not have PRs that hangs for weeks next time.

**Comment by peterpeterparker** [comment]
I'll check your comments later, thanks a lot!

**Comment by lmuntaner** [comment]
>Let's try to not have PRs that hangs for weeks next time.

smaller PRs don't hang for weeks

File: frontend/src/lib/components/accounts/CkBTCTransactionsList.svelte (hunk 3 above)
Comment (peterpeterparker) [reply frontend/src/lib/components/accounts/CkBTCTransactionsList.svelte:15]: No particular reason, can change it.

File: frontend/src/lib/components/universe/UniverseLogo.svelte (hunk 4 above)
Comment (peterpeterparker) [reply frontend/src/lib/components/universe/UniverseLogo.svelte:39]: Sure

File: frontend/src/lib/derived/universes-tokens.derived.ts (hunk 7 above)
Comment (peterpeterparker) [reply frontend/src/lib/derived/universes-tokens.derived.ts:21]: Can become a thing agree.

File: frontend/src/lib/derived/universes-tokens.derived.ts (hunk 8 above)
Comment (peterpeterparker) [reply frontend/src/lib/derived/universes-tokens.derived.ts:33]: Added a todo, makes sense

File: frontend/src/lib/stores/icrc-accounts.store.ts (hunk 10 above)
Comment (peterpeterparker) [reply frontend/src/lib/stores/icrc-accounts.store.ts:27]: Added a todo

File: frontend/src/tests/lib/components/accounts/CkBTCAccountsFooter.spec.ts (hunk 11 above)
Comment (peterpeterparker) [reply frontend/src/tests/lib/components/accounts/CkBTCAccountsFooter.spec.ts:25]: Sure

File: frontend/src/lib/derived/selected-universe.derived.ts (hunk 12 above)
Comment (peterpeterparker) [reply frontend/src/lib/derived/selected-universe.derived.ts:82]: There isn't a test for your above `selectedUniverseStore` neither, so just followed the pattern 😉

Added tests for both.

Comment (lmuntaner) [reply frontend/src/lib/derived/selected-universe.derived.ts:82]: I make mistakes 😉

**Review by lmuntaner — APPROVED** [review]
I love the feature in github that it shows only the files that have changed since last review.

It made the review much smoother!
